db/crud.go

```diff
@@ -1115,19 +1115,39 @@ func (db *Database) updateAndReturnDoc(
 	// Remove any obsolete non-winning revision bodies
 	doc.deleteRemovedRevisionBodies(db.Bucket)
 
+	// Mark affected users/roles as needing to recompute their channel access:
+	db.MarkPrincipalsChanged(docid, newRevID, changedPrincipals, changedRoleUsers)
+	return docOut, newRevID, nil
+}
+
+func (db *Database) MarkPrincipalsChanged(docid string, newRevID string, changedPrincipals, changedRoleUsers []string) {
+
+	reloadActiveUser := false
+
 	// Mark affected users/roles as needing to recompute their channel access:
 	if len(changedPrincipals) > 0 {
 		base.LogTo("Access", "Rev %q/%q invalidates channels of %s", docid, newRevID, changedPrincipals)
 		for _, name := range changedPrincipals {
 			db.invalUserOrRoleChannels(name)
-			//If this is the current in memory db.user, reload to generate updated channels
-			if db.user != nil && db.user.Name() == name {
-				user, err := db.Authenticator().GetUser(db.user.Name())
-				if err != nil {
-					base.Warn("Error reloading db.user[%s], channels list is out of date --> %+v", db.user.Name(), err)
-				} else {
-					db.user = user
+			// Check whether the active user needs to be recalculated.  Skip check if reload has already been identified
+			// as required for a previous changedPrincipal
+			if db.user != nil && reloadActiveUser == false {
+				// If role changed, check if active user has been granted the role
+				if strings.HasPrefix(name, "role:") {
+					modifiedRole := name[5:]
```

Comment:
I think this would be slightly clearer if it was refactored out to `modifiedRole := extractRole(name)`.  Also if the `name` was empty or < 5 chars long, it could return empty and log a warning rather than panicking.

Reply:
I didn't modify this originally because the strings.HasPrefix is what guarantees the safety of the slice operation.

However, I agree it makes sense to standardize both this usage, and remove the usage of the hardcoded "role:" string in general, so I've refactored this into a utility method/const in channels.